### ISSUE TITLE
Rhizome direct operations: push, pull, sync

### ISSUE THREAD

**Issue by quixotique** [servalproject/serval-dna#9]
To support Rhizome file sharing and MeshMS via infrastructure, it must be possible to configure Rhizome to connect directly to other Rhizome servers at know locations.

This issue will support a single, optional remote Rhizome instance defined by the configuration items:
- `rhizome.direct.address` a string passed to getaddrinfo(3)
- `rhizome.direct.port` a port number in decimal

Setting either of these without the other will log a warning and ignore it when any of the commands described below is invoked.

The following new servald commands will query servald for local bundles, then connect directly to the configured remote instance, send IHAVE announcements in batches to announce all locally stored bundles, then receive IHAVE and SENDME replies from the remote instance, then perform GET and/or POST operations to exchange bundles:
- **servald rhizome push** will only perform POST operations to satisfy the SENDME replies from the direct remote
- **servald rhizome pull** will only perform GET operations to fetch any bundles discovered from IHAVE announcements from the direct remote
- **servald rhizome sync** will perform a mixture of POST and GET operations, equivalent to performing a push and a pull

These commands will terminate when all bundles identified by the initial IHAVE/SENDME exchange have been transferred.  If other bundles become available during the session, these will not be detected.

**Comment by gardners** [comment]
PULL syncing will use mostly GETs, but will still need to do a POST to do the initial negotiation with the remote end to work out what needs to be PULLed.  But once that is known, then the existing GET-based requests can be used to implement that.

**Comment by quixotique** [comment]
All tests pass on Linux and Solaris except the five **rhizomeprotocol** tests relating specifically to Rhizome Direct functions:

```
bash-3.00$ ../bash-4.2/bash tests/rhizomeprotocol 
1. New bundle and update transfer to one node... PASS
2. Big new bundle transfers to one node... PASS
3. New bundle transfers to four nodes... PASS
4. Payload deletion transfers to one node... PASS
5. One way push bundle to unconnected node... FAIL
6. One way pull bundle from unconnected node... FAIL
7. Two-way sync bundles between unconnected nodes... FAIL
8. Import bundle using HTTP POST multi-part form.... FAIL
9. Add file locally using HTTP, returns manifest... FAIL
9 tests, 4 pass, 5 fail, 0 error
```

**Comment by gardners** [comment]
We now have a working solution, and just need to cherrypick in the best of both Andrew and my changes and make sure it still works.  My commits are on rhizomedirect-pgs, Andrew's on rhizomedirect, and I will attempt a merge on rhizomedirect-merge.

**Comment by quixotique** [comment]
The Rhizome Direct code branch has been merged into the master branch (89343c69cd) and remaining issues with failing tests have been resolved (as of 8b9505670d).

The Rhizome Direct code still needs a lot of improvement, such as:
- convert blocking procedural code to asynchronous code using `struct sched_ent`
- selective push/pull/sync (eg, based on bundle sender, recipient, size, date, service, etc.)
- take advantage of an improved BAR format (see issue #19)
- move push/pull into a non-blocking daemon process (see issue #1)

All of this work falls under the scope of other issue reports, so closing this issue now as done.
